### PR TITLE
Try to fix #255, #242

### PR DESCRIPTION
**Pull request by Kuniwak**
# What is changed

Before this patch, Vint can be crashed by short scripts that including Emoji such as 💜.

The cause is Chardet that is a package for encoding detection.
Chardet can be confused by short scripts. For example, Chardet returned wrong encodings by the short script that is noted on #255:

```console
$ python2 dev_tool/show_chardet_result.py test/fixture/encodings/issue-225.vim
{'confidence': 0.5658124254347925,
 'encoding': 'Windows-1254', 👈 It is wrong. The correct encoding is UTF-8.
 'language': 'Turkish'}
```

After this patch, Vint read `scriptencoding` and try to decode the encoding.

1. If the input is empty, this encoding will be `ascii`
2. Try to decode by the encoding that specified with `:scriptencoding`
3. Try to decode by the encoding that detected by Chardet
4. Try to decode as UTF-8 that is an encoding widely used
5. Finally, Vint raises `EncodeDetectionError`

```console
$ python3 ./dev_tool/show_encoding.py test/fixture/encodings/issue-225.vim
{'composed_strategies': ['DecodingStrategyForEmpty',
                         'DecodingStrategyByScriptencoding',
                         'DecodingStrategyByChardet',
                         'DecodingStrategyForUTF8'],
 'empty': 'false',
 'scriptencoding': b'utf-8',
 'selected_strategy': 'DecodingStrategyByScriptencoding',
 'version': '3.6.0 (v3.6.0:41df79263a11, Dec 22 2016, 17:23:13) \n'
            '[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]'}
```

### CODE REVIEW

**Comment by Kuniwak** [comment]
Oops, I missed your patch #226... 😱 

And I think #226 has an advantage of multiple `scriptencoding` behavior simulation.
I'm gonna merge your code first, and modify to add UTF-8 fallback.

**Comment by blueyed** [comment]
Regarding scriptencoding: it should be the preferred decoder, but it needs to handle only the chunk up to the next script encoding then.
The deciding strategies should be able maybe to indicate they not everything was decoded, but only to a certain offset?

**Comment by Kuniwak** [comment]
@blueyed You're right. My implementation is incomplete.

**Comment by Kuniwak** [comment]
Hmmm, I said that I'm gonna merge #226, but #226 did not pass test.

@blueyed would you fix #226? or I can implement the multiple scriptencoding behavior in this patch. How do you feel it?

**Comment by blueyed** [comment]
> I can implement the multiple scriptencoding behavior in this patch

That would make more sense to me.

**Comment by blueyed** [comment]
IIRC #226 was just a proof of concept / WIP, and I never came back to it.
It makes more sense to pick it from there and adopt it for your refactoring.

**Comment by Kuniwak** [comment]
OK, I'm gonna refer your code and modify in this PR.

**Comment by Kuniwak** [comment]
I implemented that simulation of multiple scriptencoding occurrence.

```console
$ cat test/fixture/encodings/multiple-scriptencoding.vim
echo "before 1"
scriptencoding utf8
echo "before 2"
scriptencoding utf8
echo "after 2"

$ python2 ./dev_tool/show_encoding.py test/fixture/encodings/multiple-scriptencoding.vim
{'0:16': {'composed_strategies': ['DecodingStrategyForEmpty',
                                  'DecodingStrategyByScriptencoding',
                                  'DecodingStrategyForUTF8',
                                  'DecodingStrategyByChardet'],
          'empty': 'false',
          'scriptencoding': 'None',
          'scriptencoding_error': '`scriptencoding` is not found',
          'selected_strategy': 'DecodingStrategyForUTF8',
          'utf-8': 'success'},
 '16:52': {'composed_strategies': ['DecodingStrategyForEmpty',
                                   'DecodingStrategyByScriptencoding',
                                   'DecodingStrategyForUTF8',
                                   'DecodingStrategyByChardet'],
           'empty': 'false',
           'scriptencoding': 'utf8',
           'selected_strategy': 'DecodingStrategyByScriptencoding'},
 '52:87': {'composed_strategies': ['DecodingStrategyForEmpty',
                                   'DecodingStrategyByScriptencoding',
                                   'DecodingStrategyForUTF8',
                                   'DecodingStrategyByChardet'],
           'empty': 'false',
           'scriptencoding': 'utf8',
           'selected_strategy': 'DecodingStrategyByScriptencoding'},
 'version': '2.7.13 (default, Feb 13 2017, 20:26:49) \n[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]'}
```

File: vint/encodings/decoder.py [hunk 1]
```diff
@@ -2,12 +2,12 @@
...
-from vint.encodings.decoding_strategy.utf8 import DecodingStrategyForUTF8
-from vint.encodings.decoding_strategy.composed import ComposedDecodingStrategy
-from vint.encodings.decoding_strategy.empty import DecodingStrategyForEmpty
-from vint.encodings.decoding_strategy.chardet import DecodingStrategyByChardet
+from vint.encodings.decoding_strategy.abstract_strategy import DecodingStrategy
+from vint.encodings.decoding_strategy.scriptencoding_strategy import DecodingStrategyByScriptencoding
+from vint.encodings.decoding_strategy.utf8_strategy import DecodingStrategyForUTF8
+from vint.encodings.decoding_strategy.composed_strategy import ComposedDecodingStrategy
+from vint.encodings.decoding_strategy.empty_strategy import DecodingStrategyForEmpty
+from vint.encodings.decoding_strategy.chardet_strategy import DecodingStrategyByChardet
```
Comment (blueyed) [inline vint/encodings/decoder.py]: What about having them all in `vint.encodings.decoding_strategy` directly?

Comment (Kuniwak) [reply vint/encodings/decoder.py]: I improved the code on ee265da

**Comment by Kuniwak** [comment]
@Coacher Could you review it? or if you have no time to review, I'm gonna merge it.

### Point of Review

- Does this code follow Python way?
- Is this code readable?

**Comment by Coacher** [comment]
@Kuniwak, will do later today. Thank you for the reminder.

**Comment by Kuniwak** [comment]
@Coacher How about it?

**Comment by Kuniwak** [comment]
Thank you for taking your time for the review!

File: vint/ast/parsing.py [hunk 2]
```diff
@@ -42,23 +34,15 @@ def parse(self, string):
...
-            decoded = bytes_seq.decode(encoding)
-            decoded_and_lf_normalized = decoded.replace('\r\n', '\n')
+        decoder = Decoder(default_decoding_strategy)
+        decoded = decoder.read(file_path)
+        decoded_and_lf_normalized = decoded.replace('\r\n', '\n')
 
+        try:
             return self.parse(decoded_and_lf_normalized)
+        except vimlparser.VimLParserException:
+            pprint(decoder.debug_hint)
```
Comment (blueyed) [inline vint/ast/parsing.py]: I think this should only get done in verbose/debug mode?

Comment (Kuniwak) [reply vint/ast/parsing.py]: Oops, it is debug code as you said.
I'm gonna remove it at next release.

**Pull request merged**